### ISSUE TITLE
[rom] What to do if reset reason is wrong?

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#7974]
This came up through the security consultants.

Right now, Mask ROM probably performs its early function based on what reset reason it thinks it sees. 
However, what if the reset reason itself is vulnerable to attacks, could this direct ROM to keep information that it should have erased? 

For example, what should have been "POR" or "Security escalation" were changed to "warm boot" or "debug requested reset".

What would be the right thing to do here?
The hardware side can make this harder and use multiple bits per indication, but I think it would also make sense for software to treat it like POR if it can't make sense of the read back.

**Comment by tjaychen** [comment]
@matutem making the reason multi-bit will definitely have a DV impact.

**Comment by cdgori** [comment]
> My preference would be to trigger shutdown in this situation rather than assuming POR.

Can you elaborate on this choice?  I'm honestly not sure whether shutdown or behave-like-POR is correct here.
(If I assume shutdown is the better option - if the shutdown completes correctly, I presume the next boot cycle would behave as POR?)

Also, multi-bit encoding of this to make FI slightly harder seems like a good idea, but would be curious to know if this has heavy DV overhead.

**Comment by mundaym** [comment]
> Can you elaborate on this choice? I'm honestly not sure whether shutdown or behave-like-POR is correct here.

If the reset reason is corrupt then presumably the chip is either faulty, in an inconsistent state or is being tampered with. In my opinion we should not continue to boot in any of those situations (at least in `PROD`). If nothing else we risk exposing 'impossible' values to other software.

What is your opinion @moidx?

**Comment by cdgori** [comment]
Agreed, something is clearly wrong if the reset reason is corrupt.  

I'm not sure whether to have more confidence in the POR sequence or the shutdown sequence as a clean-up mechanism in this case though.  It's probably true that trying to lift the system into its booted state (as if POR happened, regardless of reset reason) even in the face of this corruption could be bad, but it's also possible that the shutdown could be halted/bypassed (e.g. by withholding clocks) and then the non-cleared state might be accessible to the adversary.

Basically the shutdown state to me isn't guaranteed to be "clean" (because it's hard to really do that in the process of shutting everything off) but a non-corrupt-reset-reason POR is basically guaranteed to be clean (because all the sensors and checks are enabled), but that might be some bias in my mind.  I'm having trouble exactly knowing what the potential threats/attack paths are for a corruption-detected POR.

After noodling on this a bit more (and re-writing this comment about 4 times while we were talking about this issue in the security meeting this morning), I think I probably agree that the risk of the POR not successfully clearing out state (and thereby exposing 'impossible' values as you said) is higher than the risk of the shutdown being early-terminated, just due to what I think the complexity of the attack is, but I'm not sure how to validate that intuition.  So it's ~probably better to just do the shutdown and try to get a clean POR on the subsequent boot.

**Comment by felixmiller** [comment]
This issue was discussed in the Security WG meeting on 2021-09-02. Feedback/summary:

Tim: still need to think about protecting the register and the logic driving it but what if the content read out doesn’t make any sense?
Tim: its actually two issues: 1) reset value is wrong, 2) escalation process gets to ROM
Tim: escalation is currently persistent, 
Rupert: could we instead to some hard to fake stuff before the warm boot?
Miguel: Another open is what if we have pending escalations?
Miguel: could define a info Flash page
Rupert: probably to slow, reset can be triggered before the Flash write
Tim: can also not use retention RAM, but it can’t be trusted to not discard data during warm boot because of the ephemeral key used
tim: I’d have to take a look to see if scratch registers in AON is a possibility.
tim: I have to check the POR as I believe it also removes the power.  
tim: Need to check storage stage after reset, and what if there are any memory targets that are not updated by reset propagation (and if there should be SW controls to force a reset).

**Action items:**
Explore POR initial state properties (e.g. initial state of critical registers vs initialization).
Add a warm qualification flag in retention RAM.
Add a POR reset type escalation policy to mask ROM (OTP backed).

**Comment by msfschaffner** [comment]
Have these action items been addressed or is there anything left to be done for this issue?

**Comment by alphan** [comment]
We still need to work on this.

**Comment by moidx** [comment]
Hi @bilgiday, can you add this to the list of issues to go through as part of the ROM security assessment? Thanks

**Comment by bilgiday** [comment]
Ack, I added this one to my task list :+1:

**Comment by msfschaffner** [comment]
Any update on where this is at right now?

@moidx @johannheyszl @cfrantz ?

**Comment by felixmiller** [comment]
We discussed this in the security WG again and recommendation is still to move forward with this (meaning multi-bit encoding of the reset reason).

What we still have to decide on is the behavior when the encoding is invalid (shutdown or reset).

CC @timothytrippel

**Comment by johannheyszl** [comment]
Removing Hotlist label, since it has been discussed recently and recommended for implementation.

**Comment by a-will** [comment]
Is this still supposed to be done by M4? Has the mubi encoding of reset reasons been deprioritized?

**Comment by moidx** [comment]
Moving this to M4 so that it is part of the next triage discussion. 

@johannheyszl, @vogelpi, can you guys sync on this before the Monday triage call? 

The additional context here is that we want to understand what is the likelihood of an attacker being able to clear the escalation reset bit or setting the POR one. There is logic in the ROM now that will treat any dirty reset as non-POR, so this is something that can be used to mitigate against potential attacks. I can share more context on this on Monday.

**Comment by johannheyszl** [comment]
A summary of the discussion on this can be found [here](https://docs.google.com/document/d/1NXM47D_GiMQ5T6YzxfC6VFRLmhA0Y42RMZ5xq3qasbs/edit?usp=sharing). Suggestion is to remove the M4 milestone. @vogelpi to check please.

**Comment by vogelpi** [comment]
Thanks @johannheyszl for preparing the summary. I support your assessment and the decision to remove the M4 milestone here.

**Comment by moidx** [comment]
Discussed during triage meeting and decided not to implement MUBI encoding for `Earlgrey.PROD`. Moving this issue to the `ROM` to ensure we are checking the consistency of the reset reason. 

By consistency, we are concerned about having a clear plan wrt handling accumulated reasons in the sticky register.

Additional discussion notes are available in the opentitan.org domain.

**Comment by moidx** [comment]
This issue may impact the resolution plan for https://github.com/lowRISC/opentitan/issues/21355.

**Comment by cfrantz** [comment]
The strategy to deal with a corrupted reset reason is to double-check the register and possibly cause a boot-fault if the values are not the same.  The policy on whether or not to cause a boot-fault is controlled by an OTP configuration word.
